### PR TITLE
Use packable BTreeSet impls

### PR DESCRIPTION
**Pull request by DaughterOfMars**
# Description of change

This PR uses BTreeSets rather than boxed slices where possible when unique, ordered data is required.

# Related Issues

Blocked by https://github.com/iotaledger/common-rs/pull/70

### CODE REVIEW

**Review by thibault-martinez — CHANGES_REQUESTED** [review]
Conflicts

File: sdk/src/types/block/error.rs [hunk 1]
```diff
@@ -364,9 +365,10 @@ impl fmt::Display for Error {
             Self::UnallowedFeature { index, kind } => {
                 write!(f, "unallowed feature at index {index} with kind {kind}")
             }
-            Self::UnallowedUnlockCondition { index, kind } => {
+            Self::DisallowedUnlockCondition { index, kind } => {
                 write!(f, "unallowed unlock condition at index {index} with kind {kind}")
```
Comment (Thoralf-M) [inline sdk/src/types/block/error.rs]: Should also be renamed here then?
```suggestion
                write!(f, "disallowed unlock condition at index {index} with kind {kind}")
```

**Review by Thoralf-M — DISMISSED** [review]

**Comment by thibault-martinez** [comment]
Conflicts

**Review by Thoralf-M — DISMISSED** [review]

**Comment by thibault-martinez** [comment]
Solve conflicts and I'll review/merge today

**Review by Thoralf-M — DISMISSED** [review]

File: sdk/src/types/block/rand/parents.rs [hunk 2]
```diff
@@ -7,6 +7,9 @@ use crate::types::block::{
 };
 
 /// Generates random strong parents.
-pub fn rand_strong_parents<const MAX: u8>() -> Parents<1, MAX> {
-    Parents::from_vec(rand_block_ids(rand_number_range(Parents::<1, MAX>::COUNT_RANGE).into())).unwrap()
+pub fn rand_parents<const MIN: u8, const MAX: u8>() -> Parents<MIN, MAX> {
```
Comment (thibault-martinez) [inline sdk/src/types/block/rand/parents.rs:10]: Nice, I almost commented in the previous PR that I didn't like it and ended up not doing it

Comment (DaughterOfMars) [reply sdk/src/types/block/rand/parents.rs:10]: this comment made me notice the comment above is outdated, thanks lol

File: sdk/src/types/block/output/unlock_condition/mod.rs [hunk 3]
```diff
@@ -337,83 +363,86 @@ impl UnlockConditions {
 
     /// Creates a new [`UnlockConditions`] from a vec.
     pub fn from_vec(unlock_conditions: Vec<UnlockCondition>) -> Result<Self, Error> {
```
Comment (thibault-martinez) [inline sdk/src/types/block/output/unlock_condition/mod.rs:365]: So this wouldn't error anymore on duplicates ?

Comment (DaughterOfMars) [reply sdk/src/types/block/output/unlock_condition/mod.rs:365]: yeah currently it won't

Comment (thibault-martinez) [reply sdk/src/types/block/output/unlock_condition/mod.rs:365]: Can we somehow bring it back?

Comment (DaughterOfMars) [reply sdk/src/types/block/output/unlock_condition/mod.rs:365]: done

**Review by Thoralf-M — DISMISSED** [review]

**Review by Thoralf-M — DISMISSED** [review]

**Review by Thoralf-M — DISMISSED** [review]

**Review by Thoralf-M — APPROVED** [review]

File: sdk/src/types/block/error.rs [hunk 4]
```diff
@@ -371,9 +372,10 @@ impl fmt::Display for Error {
             Self::UnallowedFeature { index, kind } => {
                 write!(f, "unallowed feature at index {index} with kind {kind}")
             }
-            Self::UnallowedUnlockCondition { index, kind } => {
-                write!(f, "unallowed unlock condition at index {index} with kind {kind}")
+            Self::DisallowedUnlockCondition { index, kind } => {
+                write!(f, "disallowed unlock condition at index {index} with kind {kind}")
             }
+            Self::TooManyUnlockConditions => write!(f, "too many unlock conditions"),
```
Comment (kwek20) [inline sdk/src/types/block/error.rs:378]: maybe we want max/current added?

File: sdk/src/types/block/output/delegation.rs [hunk 5]
```diff
@@ -359,7 +359,7 @@ impl DelegationOutput {
 
     // Transition, just without full ValidationContext.
     pub(crate) fn transition_inner(current_state: &Self, next_state: &Self) -> Result<(), StateTransitionError> {
-        if !(current_state.delegation_id.is_null() && !next_state.delegation_id().is_null()) {
+        if !current_state.delegation_id.is_null() | next_state.delegation_id().is_null() {
```
Comment (kwek20) [inline sdk/src/types/block/output/delegation.rs:362]: nice bit stuff :D lets add a `|`

File: sdk/src/types/block/output/feature/mod.rs [hunk 6]
```diff
@@ -67,6 +71,17 @@ impl Ord for Feature {
         self.kind().cmp(&other.kind())
     }
 }
+impl core::hash::Hash for Feature {
+    fn hash<H: core::hash::Hasher>(&self, state: &mut H) {
+        self.kind().hash(state);
```
Comment (kwek20) [inline sdk/src/types/block/output/feature/mod.rs:76]: Are we sure we want the feature hash to just be the kind? that would make 2 different SenderFeatures have the same hash

Comment (kwek20) [reply sdk/src/types/block/output/feature/mod.rs:76]: just noticed that the Eq is also just kind. for our code it works, but what if someone gets outputs and compares it with their feature, youd just expect == to work.

Comment (thibault-martinez) [reply sdk/src/types/block/output/feature/mod.rs:76]: We started speaking about it on Slack, we should definitely continue the discussion. I really dislike this.

**Review by kwek20 — COMMENTED** [review]
well im not sure about all these hash/eq  only taking the kind into consideration for external people. may be confusing. idk if we actually care about that tho.

**Comment by DaughterOfMars** [comment]
I'm going to close this, as the concerns are valid and additionally I believe there is a better way to do this.

**Pull request closed without merging**